fem/restriction.cpp

```diff
@@ -0,0 +1,1351 @@
+         }
+      });
+   }
+   else
+   {
+      auto d_indices1 = scatter_indices1.Read();
+      auto d_x = Reshape(x.Read(), t?vd:ndofs, t?ndofs:vd);
+      auto d_y = Reshape(y.Write(), nd, vd, nf);
+      MFEM_FORALL(i, nfdofs,
+      {
+         const int dof = i % nd;
+         const int face = i / nd;
+         const int idx1 = d_indices1[i];
+         for (int c = 0; c < vd; ++c)
+         {
+            d_y(dof, c, face) = d_x(t?c:idx1, t?idx1:c);
+         }
+      });
+   }
+}
+
+void L2FaceRestriction::MultTranspose(const Vector& x, Vector& y) const
+{
+   // Assumes all elements have the same number of dofs
+   const int nd = dof;
+   const int vd = vdim;
+   const bool t = byvdim;
+   const int dofs = nfdofs;
+   auto d_offsets = offsets.Read();
+   auto d_indices = gather_indices.Read();
+   auto d_x = Reshape(x.Read(), nd, vd, 2, nf);
+   auto d_y = Reshape(y.Write(), t?vd:ndofs, t?ndofs:vd);
+   MFEM_FORALL(i, ndofs,
+   {
+      const int offset = d_offsets[i];
+      const int nextOffset = d_offsets[i + 1];
+      for (int c = 0; c < vd; ++c)
+      {
+         double dofValue = 0;
+         for (int j = offset; j < nextOffset; ++j)
+         {
+            int idx_j = d_indices[j];
+            bool isE1 = idx_j < dofs;
+            idx_j = isE1 ? idx_j : idx_j - dofs;
+            dofValue +=  isE1 ?
+            d_x(idx_j % nd, c, 0, idx_j / nd)
+            :d_x(idx_j % nd, c, 1, idx_j / nd);
+         }
+         d_y(t?c:i,t?i:c) += dofValue;
+      }
+   });
+   // Kept to be used with MFEM_USE_ATOMICADD
+   // if (m==L2FaceValues::Double)
+   // {
+   //    auto d_indices1 = indices1.Read();
+   //    auto d_indices2 = indices2.Read();
+   //    auto d_x = Reshape(x.Read(), nd, vd, 2, nf);
+   //    auto d_y = Reshape(y.ReadWrite(), t?vd:ndofs, t?ndofs:vd);
+   //    MFEM_FORALL(i, nfdofs,
+   //    {
+   //       const int idx1 = d_indices1[i];
+   //       const int idx2 = d_indices2[i];
+   //       for (int c = 0; c < vd; ++c)
+   //       {
+   //          MFEM_ATOMIC_ADD(d_y(t?c:idx1,t?idx1:c), d_x(i % nd, c, 0, i / nd));
+   //          if (idx2!=-1) { MFEM_ATOMIC_ADD(d_y(t?c:idx2,t?idx2:c), d_x(i % nd, c, 1, i / nd)); }
+   //       }
+   //    });
+   // }
+   // else
+   // {
+   //    auto d_indices1 = indices1.Read();
+   //    auto d_x = Reshape(x.Read(), nd, vd, nf);
+   //    auto d_y = Reshape(y.ReadWrite(), t?vd:ndofs, t?ndofs:vd);
+   //    MFEM_FORALL(i, nfdofs,
+   //    {
+   //       const int idx1 = d_indices1[i];
+   //       for (int c = 0; c < vd; ++c)
+   //       {
+   //          MFEM_ATOMIC_ADD(d_y(t?c:idx1,t?idx1:c), d_x(i % nd, c, i / nd));
+   //       }
+   //    });
+   // }
+}
+
+#ifdef MFEM_USE_MPI
+
+ParL2FaceRestriction::ParL2FaceRestriction(const ParFiniteElementSpace &fes,
+                                           ElementDofOrdering e_ordering,
+                                           FaceType type,
+                                           L2FaceValues m)
```

Comment:
I completely agree with this, unfortunately I don't think it is possible because most of the `Par` objects don't even exist in serial. Also, the code is not really duplicated, there is differences.